Development

JorritvandenBerg opened the pull request:
**What current issue(s) does this address, or what feature is it adding?**
Informing the user that an insufficient funds error is possibly caused by an unsynced wallet and how to resolve that.

**How did you solve this problem?**
Extended the error message

**How did you make sure your solution works?**
Tested

**Are there any special changes in the code that we should be aware of?**
No, just a string that has changed

**Please check the following, if applicable:**

- [ ] Did you add any tests?
- [x] Did you run `make lint`?
- [x] Did you run `make test`?
- [x] Are you making a PR to a feature branch or development rather than master? <- development
- [x] Did you add an entry to `CHANGELOG.rst`? (if not, please do)

metachris commented:
I think we could improve this PR by checking if the wallet is not synced, and only showing the error message then. What do you think about this?

JorritvandenBerg commented:
Updated, I added GetSyncState to the Wallet class to verify this and throw a warning if the wallet is not fully synced. Sorry, commits became a bit messy. I had to fiddle a bit to reproduce the case, which is not easy on a private, but on test net it often occurs. Below is a screenshot of the warning.

<img width="1073" alt="screen shot 2018-02-26 at 23 37 04" src="https://user-images.githubusercontent.com/7474467/36700182-3ed70ebe-1b4f-11e8-9edb-0d2e49a1ae0f.png">

metachris commented on neo/Wallets/Wallet.py:
```diff
@@ -1205,6 +1213,18 @@ def GetSyncedBalances(self):
+            bool: True if wallet is synced.
+        """
+        if (int(100 * self._current_height / Blockchain.Default().Height)) < 100:
+            return None
```
this should return `False`.

metachris commented on neo/Wallets/Wallet.py:
```diff
@@ -1205,6 +1213,18 @@ def GetSyncedBalances(self):
                 balances.append((asset.symbol, self.GetBalance(asset)))
         return balances
 
+    def GetSyncedState(self):
```
I would also call this method `IsSynced()`

metachris commented on neo/Wallets/Wallet.py:
```diff
@@ -1054,6 +1054,14 @@ def MakeTransaction(self,
 
         self._vin_exclude = None
 
+        synced_state = self.GetSyncedState()
```
rename this to `is_synced` pls

metachris commented on neo/Wallets/Wallet.py:
```diff
@@ -1054,6 +1054,14 @@ def MakeTransaction(self,
+        synced_state = self.GetSyncedState()
+
+        if not synced_state:
+            logger.warn("Wait for your wallet to be synced before doing \
```
Don't add line breaks, please use this syntax for long strings:
 
    s = ("this is a very"
         "long string too"
         "for sure ...")

JorritvandenBerg commented:
Ah those did not trigger on my machine. Will try with the is_synced block wrapped in the check for unspents.

Also added a check if the Blockchain.Default().Height in IsSynced is not zero (as that causes a zero division error).

________________________________________
From: Chris Hager <notifications@github.com>
Sent: Tuesday, February 27, 2018 12:20:15 PM
To: CityOfZion/neo-python
Cc: Jorrit van den Berg; Author
Subject: Re: [CityOfZion/neo-python] Development (#259)

The tests fail, please take a look and find out why, and then let's talk about solutions or you just propose one.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/CityOfZion/neo-python/pull/259#issuecomment-368855937>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AHINI5Hv3Zw1ckjqQQLWvnP8EL5O9l2Tks5tY_L_gaJpZM4ST2LP>.

metachris commented:
@JorritvandenBerg please.resolve the Changeling conflict

localhuman commented:
I guess I'd prefer `IsSynced` to be an `@property`

Also, should we use `"""` for longer strings?

metachris commented:
Yes, let's make it an `@property` 👍 

In general you like this?

Re strings, I personally think this is fine and pretty standard, probably fine as is

JorritvandenBerg commented:
Adding the property decorator and running a last check. I noticed that during an initial sync of the chain, percent_synced is at 100. So the insufficient funds error is still triggered in that case. Do we want to work on that?

metachris commented:
Hm, could add another check and a warning if sync is 100 but chain is unsynced. could also do this in nother PR though.
